### ISSUE TITLE
Is there a modification on watershed segmentation on images 

### ISSUE THREAD

**Issue by Ahmedhabashi** [scikit-image/scikit-image#4398]
[I](url
![result1](https://user-images.githubusercontent.com/58392288/72462123-30600a80-3785-11ea-8914-476744312e5a.jpg)
![SampleE_#1_26pxfor1mm](https://user-images.githubusercontent.com/58392288/72462283-9187de00-3785-11ea-8267-6cdc8b7eda12.jpg)
![SampleE_#1_26pxfor1mm_out](https://user-images.githubusercontent.com/58392288/72462329-b5e3ba80-3785-11ea-96b2-e64c85565e26.jpg)

) am new to image processing on python and my main project aims to segment the particles of soil grains included within an image after performing edge detection of the image using holistically nested edge detection then applying watershed segmentation on this image to get the contour areas in the image however I face some problems regarding segmenting the intersecting grains as I got two or more grains as one grain and this gives me an error in the results . Could anyone figure out where can I modify the code to get better segmenting the grains in the image .

## Way to reproduce
```python
import cv2
import numpy as np
from skimage.feature import peak_local_max
from skimage.morphology import watershed
from scipy import ndimage
import matplotlib.pyplot as plt

# Load in image, convert to gray scale, and Otsu's threshold
image1 = cv2.imread('SampleH2_#1_69pxfor1cm.jpg')
image = cv2.pyrMeanShiftFiltering(image1,15,30)
kernel = cv2.getStructuringElement(cv2.MORPH_ELLIPSE,(5,5))
image = cv2.morphologyEx(image,cv2.MORPH_CLOSE,kernel)
gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
thresh = cv2.threshold(gray, 0, 255, cv2.THRESH_BINARY_INV + cv2.THRESH_OTSU)[1]
distance_map = ndimage.distance_transform_edt(thresh)
local_max = peak_local_max(distance_map, indices=False, footprint=np.ones((51, 51)), labels=thresh)

# Perform connected component analysis then apply Watershed
markers = ndimage.label(local_max, structure=np.ones((3, 3)))[0]
labels = watershed(-distance_map, markers, mask=thresh)

# Iterate through unique labels
n = 0
total_area = []
for label in np.unique(labels):
    if label == 0:
        continue

    # Create a mask
    mask = np.zeros(gray.shape, dtype="uint8")
    mask[labels == label] = 255

    # Find contours and determine contour area
    cnts = cv2.findContours(mask.copy(), cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
    cnts = cnts[0] if len(cnts) == 2 else cnts[1]
    c = max(cnts, key=cv2.contourArea)
    area = cv2.contourArea(c)
    total_area.append(area)
    n+=1
    print('Grain number' , ' of number ' , n , ' has area = ', area)
    # total_area += area
    cv2.drawContours(image, [c], -1, (36,255,12), 3)
# cal_area = []
# for a in total_area:
#     b= a/(64*64)
#     cal_area.append(b)


cv2.imwrite('result.jpg' ,image)
# print(total_area)
cv2.imshow('image', image)
cv2.waitKey(0)
# Place the full code we need to recreate your issue here
# upload all necessary images to github too!
```

## Version information
```python
# Paste the output of the following python commands
the code gives the areas of contours detected as shown in the attached images 

> .

```

```python
# your output here

```

**Comment by emmanuelle** [comment]
Well, to be honest this is not so bad :-). Using the watershed on a mixture of large and small objects is usually a nightmare. You will never have a perfect segmentation for this kind of images, but you can probably improve it if you know the metric you are optimizing. Are you either:
- trying to detect a representative fraction of the grains, so that you can afford not detecting some but you want to avoid having a bias (that is, you don't want to have a bias for not detecting small grains or big grains)
- trying to detect as many grains as possible, but you can live with some bias?

Also, do you have just one image to segment, so that some manual pre-processing or post-processing (for merging regions) is acceptable, or do you need to process automatically tens to hundreds of such images?

**Comment by Ahmedhabashi** [comment]
Thank you so much for your reply ... really my main aim in this project is the second option which is I am  trying to detect as many grains as possible but I can live with some bias so as after getting the areas of contours included in an image I will draw a curve between equivalent diameters of grains detected in the image and percentage passing of grains then comparing this curve to the results of traditional sieving in the lab on the soil sample being captured in the image .. I wish and hope to be the same curve so as to introduce another contribution to the estimation of this curve for soil samples.
Furthermore , I have  hundreds of images of different soil samples mixed in the lab & I hope to get the processing automatically .. so I hope that you could help me to the way I can enhance the segmentation of the grains using this procedure .. Many thanks again .

**Comment by emmanuelle** [comment]
Hi @Ahmedhabashi ok, so there are several possible approaches.
- You can keep the same approach and try to do automatic post-processing on your labels, ie have a criterion for merging neighbor labels (the criterion might be that the average norm of the gradient if smaller than a threshold on the boundary between the two labels, or a composite criterion including also length and shape criteria on the boundary, etc.). 
- It might be good to do a manual (supervized) post-processing by merging yourself the labels which need to be merged, so that you can measure the accuracy on your automatic method. A while ago I developed a tool for merging labels based on Dash, it's on https://dash-canvas.plotly.host/correct-segmentation/ (you need to split a label first and then you can merge it). The source code is https://github.com/plotly/canvas-portal/blob/master/apps/correct-segmentation/app.py you will need to adapt it slightly if you want for example to save the updated segmentation as a numpy array file. 
- If you want a really good segmentation and you're ready to invest some time in it, you can try an approach based on deep learning (there are possible approaches for this, one is to classify pixels as 0 and 1 whether they lie on boundaries between grains, more complex approaches can try to segment directly different grains). I'm not a specialist of this so I would recommend to explore the literature on cell segmentation in biology because similar problems are encountered. You could for example train a CNN (https://keras.io/examples/mnist_cnn/). For this you will also need a ground truth, so the label-merging tool can be handy. Some data augmentation methods (https://imgaug.readthedocs.io/en/latest/) also help if you cannot have a lot of labels in your ground truth.

**Comment by imagejan** [comment]
This issue is a usage question much more than an issue with the `scikit-image` code base. I believe other people with similar image analysis problems would highly benefit if this kind of discussions would happen on the forum (https://forum.image.sc) instead of here in a GitHub issue. The forum offers a wider visibility and a better searchability (in my opinion).

What do others think? How about directing authors of posts like this one right-away to the forum, instead of maintaining the `type: support` label here?

**Comment by emmanuelle** [comment]
@imagejan I tend to agree with you. We sort of have a chicken and egg problem with the forum because it's not a scikit-image specific forum and there are hardly any topic tagged with scikit-image, so maybe users don't feel like posting on it, I don't know. We could modify the issue template saying "If you're looking for guidance on how to use scikit-image, please post of the forum...". What do you think? Would you like to take care of this @imagejan ?

**Comment by imagejan** [comment]
> We could modify the issue template saying "If you're looking for guidance on how to use scikit-image, please post of the forum...". What do you think?

👍 I opened https://github.com/scikit-image/scikit-image/pull/4429 for this.

**Comment by Ahmedhabashi** [comment]
@emmanuelle  Thank you so much for your kind reply and support . However , I would prefer going on your first approach but I don't have the enough experience to modify the labels in my algorithm for better segmentation however I found that (min distance ) in peak_local_max function has a big influence on the segmentation process .Furthermore , I am trying to modify the labels but not able to find a significant difference so I hope that you could help me with methods to modify the code for better automated segmentation , I would be very grateful to you , as I am civil engineer and I am new to this image processing field
Here is attached another pictures for watershed segmentation and the defects in detection of large grains as it divides it into 3 or 4 parts which mislead the results . 
![Sample12_#2_71pxfor1cmm](https://user-images.githubusercontent.com/58392288/73545828-ffbcd980-43f0-11ea-802a-4fab7806aaa5.jpg)
![Sample12_#2_71pxfor1cmm_out](https://user-images.githubusercontent.com/58392288/73545829-00557000-43f1-11ea-8878-628fe430fa93.jpg)
![Sample12_#2_71pxfor1cmm_outt](https://user-images.githubusercontent.com/58392288/73545830-00557000-43f1-11ea-8fe9-296d569ef4e8.jpg)

**Comment by Ahmedhabashi** [comment]
@emmanuelle  Do you have any recommendation for me  regarding the last comment , I would be so grateful

**Comment by soupault** [comment]
Closing as not an issue, but, please, feel free to continue the discussion here or ask from an image processing community at https://forum.image.sc/.

**Comment by Ostyk** [comment]
Hi, @Ahmedhabashi 

I wanted to ask which network did you use for the boundary detection. I'm facing a similar task but with dense leaves instead of soil grains, like in your case. Also did you manage to find a way to merge these segments effectively?
